### PR TITLE
Migrate brave-core changes from Android Ads

### PR DESCRIPTION
**Pull request by tmancey**
Fixes https://github.com/brave/brave-browser/issues/5160

## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [ ] Added/updated tests for this change (for new code or code which already has tests).
- Verified that these changes build without errors on
  - [ ] Android
  - [x] iOS
  - [ ] Linux
  - [x] macOS
  - [ ] Windows
- Verified that these changes pass automated tests (unit, browser, security tests) on
  - [x] iOS
  - [ ] Linux
  - [x] macOS
  - [ ] Windows
- [x] Verified that all lint errors/warnings are resolved (`npm run lint`)
- [x] Ran `git rebase master` (if needed).
- [x] Ran `git rebase -i` to squash commits (if needed).
- [x] Tagged reviewers and labelled the pull request as needed.
- [ ] Request a security/privacy review as needed.
- [x] Add appropriate QA labels (QA/Yes or QA/No) to include the closed issue in milestone
- [ ] Public documentation has been updated as necessary. For instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protection-Mode
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-compatibility-issues-with-tracking-protection
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide

## Test Plan:

- Confirm ad notifications are shown
- Switch ads off and then back on and confirm ad notifications are still shown (when turning off ads, state is removed so you will need to visit a website for page classification)
- Confirm ads are cashed-in
- Confirm clicking an ad opens in a new tab (or same tab for duplicate ad)
- Confirm `default/ads_service` state is removed when ads are switched off
- Confirm `default/ads_service` state is created when ads are switched on

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions 
- [ ] Verify test plan is specified in PR before merging to source

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on.
- [ ] All relevant documentation has been updated.

### CODE REVIEW

File: vendor/bat-native-ads/include/bat/ads/notification_info.h [hunk 1]
```diff
@@ -16,14 +16,15 @@ namespace ads {
...
-  explicit NotificationInfo(const NotificationInfo& info);
+  NotificationInfo(const NotificationInfo& info);
   ~NotificationInfo();
 
   const std::string ToJson() const;
   Result FromJson(
       const std::string& json,
       std::string* error_description = nullptr);
 
+  std::string id;
```
Comment (kylehickinson) [inline vendor/bat-native-ads/include/bat/ads/notification_info.h]: Should add this property to `BATAdsNotification`

File: vendor/bat-native-ads/include/bat/ads/ads.h [hunk 2]
```diff
@@ -44,75 +51,77 @@ class ADS_EXPORT Ads {
   // Should be called to get the region for the specified locale
   static std::string GetRegion(const std::string& locale);
 
-  // Should be called when Ads are enabled or disabled on the Client
-  virtual void Initialize() = 0;
+  // Should be called when Ads is enabled on the Client
+  virtual void Initialize(InitializeCallback callback) = 0;
```
Comment (kylehickinson) [inline vendor/bat-native-ads/include/bat/ads/ads.h]: You changed the `Initialize` signature but I don't see a change to how its called in `BATBraveAds.mm` as seen here:

https://github.com/brave/brave-core/blob/40e6df39e1903e1240f54d97fe2d9a3168420c70/vendor/brave-ios/Ads/BATBraveAds.mm#L70

File: vendor/bat-native-ads/include/bat/ads/ads.h [hunk 3]
```diff
@@ -44,75 +51,77 @@ class ADS_EXPORT Ads {
...
 
-  // Should be called when Ads are enabled or disabled on the Client
-  virtual void Initialize() = 0;
+  // Should be called when Ads is enabled on the Client
+  virtual void Initialize(InitializeCallback callback) = 0;
 
-  // Should be called when the browser enters the foreground
-  virtual void OnForeground() = 0;
+  // Should be called when Ads is disabled on the Client
+  virtual void Shutdown(ShutdownCallback callback) = 0;
```
Comment (kylehickinson) [inline vendor/bat-native-ads/include/bat/ads/ads.h]: We will want to add this call here:

https://github.com/brave/brave-core/blob/40e6df39e1903e1240f54d97fe2d9a3168420c70/vendor/brave-ios/Ads/BATBraveAds.mm#L110-L117

**Comment by tmancey** [comment]
Closing as #5160 is not picking up changes

**Pull request merged**